### PR TITLE
Multithreading

### PR DESCRIPTION
**Pull request by yanzunh**
still need to add testing

### CODE REVIEW

**Comment by echen94** [comment]
Nice work on implementing multi-threading on the webserver. Stylistically, it looks good! We will need to think of a way to test this in our next step, either with unit tests or with an integration test.

**Comment by yanzunh** [comment]
Yeah i still need to make sure it is actually running concurrently instead of processing one by one. I might do it by using python to send an incomplete request first, and then send another request. After that i put the client to sleep for like 10 seconds, and finally complete the first request.

File: integration_test.py [hunk 1]
```diff
@@ -43,22 +51,43 @@ def testCase3():
     testFile.close()
     testCasesGenerator("/static/b.txt", "GET", "abcde", 3)
     os.remove(basePath + "/b.txt")
+
```
Comment (echen94) [inline integration_test.py:57]: Can add a comment explaining what multiThreadingTestCase does

**Comment by neosbob** [comment]
Good idea on how to test concurrency. Everything looks like it should work, and the integration test can help us test out this multithreading feature.

File: integration_test.py [hunk 2]
```diff
@@ -43,22 +51,43 @@ def testCase3():
...
+    response = connection.getresponse()
+    body = response.read()
+    connection.close()
+    connection2.close()
+    subprocess.call(["fuser", "-k", str(PORT) + "/tcp"])
+    serverProcess.terminate()
+    if body == "GET /echo HTTP/1.1\r\nHost: localhost:8080\r\nAccept-Encoding: identity\r\n\r\n" and body2 == "File Not found or No Handlers to handle uri.":
+        print "Multi-threading test case passed!"
+    else:
+        print "Error: multi-threading test case is not passed!"
```
Comment (echen94) [inline integration_test.py:84]: Can add a comment here explaining what main does

**Review by echen94 — COMMENTED** [review]
Nice implementation for integration test for multi-threading. Coding style looks great, and the content looks great as well. A recommendation is to add some comments explaining what the test cases do.

File: integration_test.py (hunk 1 above)
Comment (yanzunh) [reply integration_test.py:57]: yeah you are right. Totally forgot to add comments lol

**Comment by neosbob** [comment]
My comment was deleted :(. Anyways, everything looks good to go!

**Comment by echen94** [comment]
Nicely added comments for test cases. Good to merge now!

**Pull request merged**